File: go/pkg/mechbase/maintnode.go
package mechbase

import (
	"context"
	"fmt"
	"io"
	"net/http"
)

// MaintNode is the device-facing client for the maintnode integration API.
// It sends no Authorization header; the node is identified by nodeUUID.
type MaintNode struct {
	client   *Client
	nodeUUID string
}

// NewMaintNode creates a tokenless maintnode client. Empty baseURL uses DefaultBaseURL.
func NewMaintNode(nodeUUID, baseURL string, opts ...Option) *MaintNode {
	return &MaintNode{client: New("", baseURL, opts...), nodeUUID: nodeUUID}
}

// MaintNodeMeasurementInput is one item in a maintnode measurement push.
type MaintNodeMeasurementInput struct {
	MappingUUID string         `json:"mapping_uuid"`
	Data        map[string]any `json:"data"`
	Timestamp   *string        `json:"timestamp,omitempty"`
	LocalUUID   string         `json:"local_uuid,omitempty"`
}

func (n *MaintNode) path(suffix string) string {
	return fmt.Sprintf("/api/maintnode/%s%s", n.nodeUUID, suffix)
}

// Config returns the raw node config blob.
func (n *MaintNode) Config(ctx context.Context) (string, error) {
	var out string
	if err := n.client.doJSON(ctx, http.MethodGet, n.path("/config/"), nil, nil, &out); err != nil {
		return "", err
	}
	return out, nil
}

// Heartbeat reports node liveness and returns the server's ack object.
func (n *MaintNode) Heartbeat(ctx context.Context, balenaUUID string, services map[string]any) (map[string]any, error) {
	body := map[string]any{"balena_uuid": balenaUUID, "services": services}
	var out map[string]any
	if err := n.client.doJSON(ctx, http.MethodPost, n.path("/heartbeat/"), nil, body, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// PushMeasurements submits measurements collected by the node.
func (n *MaintNode) PushMeasurements(ctx context.Context, items []MaintNodeMeasurementInput) ([]MeasurementResult, error) {
	body := map[string]any{"measurements": items}
	var out []MeasurementResult
	if err := n.client.doJSON(ctx, http.MethodPost, n.path("/measurements/"), nil, body, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// AddMeasurementFile attaches a file to a node measurement (multipart).
func (n *MaintNode) AddMeasurementFile(ctx context.Context, measurementUUID, filename string, file io.Reader) (map[string]any, error) {
	if filename == "" {
		filename = "attachment"
	}
	files := []filePart{{Field: "file", Filename: filename, Reader: file}}
	var out map[string]any
	if err := n.client.doMultipart(ctx, n.path(fmt.Sprintf("/measurements/%s/file/", measurementUUID)), nil, files, &out); err != nil {
		return nil, err
	}
	return out, nil
}

File: python/src/mechbase/maintnode.py
from __future__ import annotations

import io
from pathlib import Path
from typing import Any, BinaryIO

from ._http import HttpClient
from .models import MeasurementResult


class MaintNode:
    """Device-facing client for the maintnode integration API.

    Unauthenticated; the node is identified by ``node_uuid`` in the path.

    Usage::

        node = MaintNode(node_uuid="...")  # defaults to https://app.mechbase.io
        cfg = node.config()
        node.push_measurements([{"mapping_uuid": "...", "data": {"rms": 2.3}}])
    """

    DEFAULT_BASE_URL = "https://app.mechbase.io"

    def __init__(self, *, node_uuid: str, base_url: str = DEFAULT_BASE_URL, timeout: float = 30.0):
        self._http = HttpClient(token=None, base_url=base_url, timeout=timeout)
        self.node_uuid = node_uuid

    def _path(self, suffix: str) -> str:
        return f"/api/maintnode/{self.node_uuid}{suffix}"

    def config(self) -> str:
        return self._http.request("GET", self._path("/config/"))

    def heartbeat(self, *, balena_uuid: str = "", services: dict | None = None) -> dict:
        return self._http.request(
            "POST", self._path("/heartbeat/"),
            json={"balena_uuid": balena_uuid, "services": services or {}},
        )

    def push_measurements(self, items: list[dict]) -> list[MeasurementResult]:
        body = self._http.request(
            "POST", self._path("/measurements/"), json={"measurements": list(items)}
        )
        return [MeasurementResult.from_dict(r) for r in body]

    def add_measurement_file(
        self, measurement_uuid: str, *,
        file: str | Path | BinaryIO | bytes, filename: str = "attachment",
    ) -> dict:
        fh: Any = file
        opened = False
        if isinstance(file, (str, Path)):
            filename = Path(file).name
            fh = open(file, "rb")  # noqa: SIM115
            opened = True
        elif isinstance(file, (bytes, bytearray)):
            fh = io.BytesIO(file)
        try:
            return self._http.request(
                "POST", self._path(f"/measurements/{measurement_uuid}/file/"),
                files={"file": (filename, fh)},
            )
        finally:
            if opened:
                fh.close()

    def close(self) -> None:
        self._http.close()

    def __enter__(self):
        return self

    def __exit__(self, *a):
        self.close()
